src/storage/data_object_storage_registry.rs

```diff
@@ -0,0 +1,150 @@
+use crate::storage::data_directory::Trait as DDTrait;
+
+pub trait Trait: timestamp::Trait + system::Trait + DDTrait + MaybeDebug {
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+
+    type DataObjectStorageRelationshipId: Parameter + Member + SimpleArithmetic + Codec + Default + Copy
+        + As<usize> + As<u64> + MaybeSerializeDebug + PartialEq;
+
+    type IsActiveMember: IsActiveMember<Self>;
+    type ContentIdExists: ContentIdExists<Self>;
+}
+
+static MSG_CID_NOT_FOUND: &str = "Content with this ID not found!";
+static MSG_DOSR_NOT_FOUND: &str = "No data object storage relationship found for this ID.";
+static MSG_ONLY_STORAGE_PROVIDER_MAY_CLAIM_READY: &str = "Only the storage provider in a DOSR can decide whether they're ready.";
+
+const DEFAULT_FIRST_DATA_OBJECT_STORAGE_RELATIONSHIP_ID: u64 = 1;
+
+#[derive(Clone, Encode, Decode, PartialEq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct DataObjectStorageRelationship<T: Trait> {
+    pub content_id: <T as DDTrait>::ContentId,
+    pub storage_provider: T::AccountId,
+    pub ready: bool,
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as DataObjectStorageRegistry {
+        // Start at this value
+        pub FirstDataObjectStorageRelationshipId get(first_data_object_storage_relationship_id) config(first_data_object_storage_relationship_id): T::DataObjectStorageRelationshipId = T::DataObjectStorageRelationshipId::sa(DEFAULT_FIRST_DATA_OBJECT_STORAGE_RELATIONSHIP_ID);
+
+        // Increment
+        pub NextDataObjectStorageRelationshipId get(next_data_object_storage_relationship_id) build(|config: &GenesisConfig<T>| config.first_data_object_storage_relationship_id): T::DataObjectStorageRelationshipId = T::DataObjectStorageRelationshipId::sa(DEFAULT_FIRST_DATA_OBJECT_STORAGE_RELATIONSHIP_ID);
+
+        // Mapping of Data object types
+        pub DataObjectStorageRelationshipMap get(data_object_storage_relationship): map T::DataObjectStorageRelationshipId => Option<DataObjectStorageRelationship<T>>;
+
+        // Keep a list of storage relationships per CID
+        pub DataObjectStorageRelationships get(data_object_storage_relationships): map T::ContentId => Vec<T::DataObjectStorageRelationshipId>;
+    }
+}
+
+decl_event! {
+    pub enum Event<T> where
+        <T as DDTrait>::ContentId,
+        <T as Trait>::DataObjectStorageRelationshipId,
+        <T as system::Trait>::AccountId
+    {
+        DataObjectStorageRelationshipAdded(DataObjectStorageRelationshipId, ContentId, AccountId),
+        DataObjectStorageRelationshipReadyUpdated(DataObjectStorageRelationshipId, bool),
+    }
+}
+
+impl<T: Trait> ContentHasStorage<T> for Module<T> {
+    fn has_storage_provider(which: &T::ContentId) -> bool {
+        let dosr_list = Self::data_object_storage_relationships(which);
+        return dosr_list.iter().any(|&dosr_id| Self::data_object_storage_relationship(dosr_id).unwrap().ready);
```

Comment:
unwrap() without checking value is Some will be very bad. But if that scenario occurs it means the two maps: DataObjectStorageRelationshipMap, DataObjectStorageRelationships were inconsistently updated. The question is how do want to fail?

Reply:
Well, the question for me is: is it possible for someone to modify the list other than through the function provided in the module?

Because if that's not possible, then these two storage items will be kept in sync, and `unwrap()` will always succeed.

I completely understand your point, for what it's worth, I just decided this was a case where failing terribly might actually be the better choice, because there must be some kind of unforeseen circumstance leading to this (i.e. likely a bug).

WDYT?

Reply:
> is it possible for someone to modify the list other than through the function provided in the module?
You never know how your code will be used in the future. It can be a case that somebody will not pay attention and set an option to `None`.

Reply:
Yeah, that has not been my best experience in the past, but ok. I can change it.

Reply:
Unfortunately, there's no elegant one liner here. Returning a default constructed value via `unwrap_or_else` or something like that is just going to be wasteful, and not more readable.